cgen/__init__.py

```diff
@@ -40,124 +40,145 @@
         raise ValueError(f"unable to map dtype '{dtype}'")
 
 
 class Generable(ABC):
+    mapper_method: ClassVar[str]
+
     def __str__(self) -> str:
-        """Return a single string (possibly containing newlines) representing
-        this code construct."""
+        """
+        :returns: a single string (possibly containing newlines) representing
+            this code construct.
+        """
         return "\n".join(line.rstrip() for line in self.generate())
 
     @abstractmethod
-    def generate(self, with_semicolon: bool = True) -> Generator[str]:
+    def generate(self, with_semicolon: bool = True) -> Iterator[str]:
         """Generate (i.e. yield) the lines making up this code construct."""
 
 
 # {{{ declarators
 
+DeclPair: TypeAlias = tuple[list[str], str]
+
+
 class Declarator(Generable, ABC):
-    def generate(self, with_semicolon: bool = True) -> Generator[str]:
+    def generate(self, with_semicolon: bool = True) -> Iterator[str]:
         tp_lines, tp_decl = self.get_decl_pair()
         tp_lines = list(tp_lines)
+
         yield from tp_lines[:-1]
-        sc = ";"
-        if not with_semicolon:
-            sc = ""
+
+        sc = ";" if with_semicolon else ""
         if tp_decl is None:
             yield f"{tp_lines[-1]}{sc}"
         else:
             yield f"{tp_lines[-1]} {tp_decl}{sc}"
 
     @abstractmethod
-    def get_decl_pair(self) -> tuple[Sequence[str], str]:
-        """Return a tuple ``(type_lines, rhs)``.
-
-        *type_lines* is a non-empty list of lines (most often just a
-        single one) describing the type of this declarator. *rhs* is the right-
-        hand side that actually contains the function/array/constness notation
-        making up the bulk of the declarator syntax.
+    def get_decl_pair(self) -> DeclPair:
+        """
+        :returns: a tuple ``(type_lines, rhs)``. *type_lines* is a non-empty list
+            of lines (most often just a single one) describing the type of this
+            declarator. *rhs* is the right-hand side that actually contains the
+            function/array/constness notation making up the bulk of the
+            declarator syntax.
         """
+
+    def struct_format(self) -> str:
+        raise NotImplementedError
+
+    def alignment_requirement(self) -> int:
+        raise NotImplementedError
+
+    def struct_maker_code(self, data: str) -> str:
+        raise NotImplementedError
+
+    def default_value(self) -> Any:
         raise NotImplementedError
 
     def inline(self, with_semicolon: bool = True) -> str:
-        """Return the declarator as a single line."""
+        """
+        :returns: the declarator as a single line.
+        """
         tp_lines, tp_decl = self.get_decl_pair()
-        tp_lines = " ".join(tp_lines)
+        tp = " ".join(tp_lines)
+
         if tp_decl is None:
-            return tp_lines
+            return tp
         else:
-            return f"{tp_lines} {tp_decl}"
+            return f"{tp} {tp_decl}"
```

Comment:
This also doesn't seem to use the `with_semicolon`. Is that ok?

Reply:
The semicolon use in various places is pretty spotty from what I've noticed. Maybe the best thing to do at the moment is not change it, so that nothing breaks :(